blockchain/utxocache.go

```diff
@@ -0,0 +1,1085 @@
+	// This method exists due to the legacy spend journal database structure.
+	// Its execution is very inefficient, but it's almost never used.
+	getEntryByHash(hash *chainhash.Hash) (*UtxoEntry, error)
+}
+
+// utxoBatcher is an interface that allows a caller to batch fetch UTXOs from
+// an underlying view.
+type utxoBatcher interface {
+	utxoView
+
+	// getEntries attempts to fetch a serise of entries from the UTXO view.
+	// If a single entry is not found within the view, then an error is
+	// returned.
+	getEntries(outpoints []wire.OutPoint) ([]*UtxoEntry, error)
+}
+
+// mapSlice is a slice of maps for Utxo entries.  The slice of maps are needed to
+// guarantee that the map will only take up N amount of bytes.  As of v1.20, the
+// go runtime will allocate 2^N + few extra buckets, meaning that for large N, we'll
+// allocate a lot of extra memory if the amount of entries goes over the previously
+// allocated buckets.  A slice of maps allows us to have a better control of how much
+// total memory gets allocated by all the maps.
+type mapSlice struct {
+	maps []map[wire.OutPoint]*UtxoEntry
+
+	// maxEntries is the maximum amount of elemnts that the map is allocated for.
+	maxEntries []int
+
+	// maxTotalMemoryUsage is the maximum memory usage in bytes that the state
+	// should contain in normal circumstances.
+	maxTotalMemoryUsage uint64
+}
+
+// length returns the length of all the maps in the map slice added together.
+func (ms *mapSlice) length() int {
+	var l int
+	for _, m := range ms.maps {
+		l += len(m)
+	}
+
+	return l
+}
+
+// size returns the size of all the maps in the map slice added together.
+func (ms *mapSlice) size() int {
+	var size int
+	for _, num := range ms.maxEntries {
+		size += CalculateRoughMapSize(num, bucketSize)
+	}
+
+	return size
+}
+
+// get looks for the outpoint in all the maps in the map slice and returns
+// the entry.  nil and false is returned if the outpoint is not found.
+func (ms *mapSlice) get(op wire.OutPoint) (*UtxoEntry, bool) {
+	var entry *UtxoEntry
+	var found bool
+
+	for _, m := range ms.maps {
+		entry, found = m[op]
+		if found {
+			break
+		}
+	}
+
+	return entry, found
+}
+
+// put puts the outpoint and the entry into one of the maps in the map slice.  If the
+// existing maps are all full, it will allocate a new map based on how much memory we
+// have left over.  Leftover memory is calculated as:
+// maxTotalMemoryUsage - (totalEntryMemory + mapSlice.size())
+func (ms *mapSlice) put(op wire.OutPoint, entry *UtxoEntry, totalEntryMemory uint64) {
+	var success bool
+	for i, maxNum := range ms.maxEntries {
+		m := ms.maps[i]
+		_, found := m[op]
+		if found {
+			// If the key is found, overwrite it.
+			m[op] = entry
+			success = true
+			continue
+		}
+		if len(m) >= maxNum {
+			// Don't try to insert if the map already at max since
+			// that'll force the map to allocate double the memory it's
+			// currently taking up.
+			continue
+		}
+
+		m[op] = entry
```

Comment:
looks like this should be checking success first, otherwise it will insert it in all of the maps? Alternatively just return when it is first found.

Reply:
Ah good point. I'll change the code to return when the key is found and write a test for it.

I guess this wasn't being triggered since you'd never insert duplicate outpoints.